spec/models/allocation_spec.rb

```diff
@@ -27,4 +31,80 @@
       expect(allocation.versions.map(&:event)).to eq(%w[create])
     end
   end
+
+  describe 'cancellation_reason' do
+    context 'when the allocation is not cancelled' do
+      let(:allocation) { build(:allocation, status: nil) }
+
+      it { expect(allocation).to validate_absence_of(:cancellation_reason) }
+    end
+
+    context 'when the allocation is cancelled' do
+      let(:allocation) { build(:allocation, status: 'cancelled') }
+
+      it {
+        expect(allocation).to validate_inclusion_of(:cancellation_reason)
+          .in_array(%w[
+            made_in_error
+            supplier_declined_to_move
+            other
+            lack_of_space_at_receiving_establishment
+            sending_establishment_failed_to_fill_allocation
+          ])
+      }
+    end
+  end
+
+  describe '#cancel' do
+    let(:allocation) { create(:allocation, :with_moves, status: nil) }
+
+    it 'changes the status of an allocation to cancelled' do
+      allocation.reload.cancel
+
+      expect(allocation.reload.status).to eq(described_class::ALLOCATION_STATUS_CANCELLED)
+    end
+
+    it 'changes the status of all associated moves to cancelled' do
+      allocation.reload.cancel
+
+      expect(allocation.reload.moves.pluck(:status)).to contain_exactly(Move::MOVE_STATUS_CANCELLED)
+    end
+
+    it 'sets the cancellation reason to other' do
+      allocation.reload.cancel
+
+      expect(allocation.reload.cancellation_reason).to eq(described_class::CANCELLATION_REASON_OTHER)
+    end
+
+    it 'sets the cancellation reason comment to cancelled by allocation' do
+      allocation.reload.cancel
+
+      expect(allocation.reload.cancellation_reason_comment).to eq('Allocation was cancelled')
+    end
+
+    it 'sets the cancellation reason on moves to other' do
+      allocation.reload.cancel
+
+      expect(allocation.reload.moves.first.cancellation_reason).to eq(Move::CANCELLATION_REASON_OTHER)
+    end
+
+    it 'sets the cancellation reason comment on moves to cancelled by allocation' do
+      allocation.reload.cancel
+
+      expect(allocation.reload.moves.first.cancellation_reason_comment).to eq('Allocation was cancelled')
+    end
+
+    it 'throws validation error if allocation invalid' do
+      allocation.from_location = nil
+
+      expect { allocation.cancel }.to raise_error(ActiveRecord::RecordInvalid)
+    end
+
+    it 'does not update moves if allocation invalid' do
+      allocation.from_location = nil
+      allocation.cancel
+    rescue ActiveRecord::RecordInvalid
```

Comment:
this test is dangerous: if `allocation.cancel` does not raise an exception, then there is no expectation and this will will pass. 

Instead test that an exception is raised, and also that the record has the correct state.

Reply:
oh, sorry, I overlooked the test above which is doing that. instead, allow allocation to raise and catch, and put the expectation in the main block:

```
...
allocation.cancel rescue ActiveRecord::RecordInvalid
expect(foo).to eql (bla)
```